File: plugins/agent-skills/skills/claude-delegate/scripts/claude_delegate/transport.py
"""Transport utilities for stream-json events.

``parse_transport`` and ``extract_tool_uses`` have been superseded by
the SDK adapter in ``sdk_transport.py``.  This module retains only
``summarize_tool_uses`` which is used by the pipeline to compact
tool-use records in the normalised envelope.
"""
from __future__ import annotations


def _truncate_text(value: str, *, limit: int = 160) -> tuple[str, bool]:
    normalized = value.replace("\r\n", "\n")
    if len(normalized) <= limit:
        return normalized, False
    return normalized[: limit - 3] + "...", True


def _summarize_input_field(key: str, value: object) -> dict[str, object]:
    if isinstance(value, str):
        preview, truncated = _truncate_text(value)
        if not truncated:
            return {key: preview}
        return {
            f"{key}_preview": preview,
            f"{key}_length": len(value),
            f"{key}_truncated": True,
        }
    if isinstance(value, (int, float, bool)) or value is None:
        return {key: value}
    if isinstance(value, list):
        sample: list[object] = []
        for item in value[:3]:
            if isinstance(item, str):
                sample.append(_truncate_text(item, limit=80)[0])
            elif isinstance(item, (int, float, bool)) or item is None:
                sample.append(item)
            elif isinstance(item, dict):
                sample.append({"keys": sorted(item.keys())[:8]})
            else:
                sample.append(type(item).__name__)
        summary: dict[str, object] = {f"{key}_count": len(value)}
        if sample:
            summary[f"{key}_sample"] = sample
        return summary
    if isinstance(value, dict):
        return {f"{key}_keys": sorted(value.keys())}
    return {f"{key}_type": type(value).__name__}


def summarize_tool_uses(tool_uses: list[dict]) -> list[dict]:
    summarized: list[dict] = []
    for item in tool_uses:
        tool_input = item.get("input")
        summary = {
            "id": item.get("id"),
            "name": item.get("name"),
            "input": {},
        }
        if isinstance(tool_input, dict):
            summarized_input: dict[str, object] = {}
            for key, value in tool_input.items():
                summarized_input.update(_summarize_input_field(key, value))
            summary["input"] = summarized_input
        else:
            summary["input"] = _summarize_input_field("value", tool_input)
        summarized.append(summary)
    return summarized

File: plugins/agent-skills/skills/claude-delegate/src/lib/transport.ts
function truncateText(
  value: string,
  limit = 160,
): [string, boolean] {
  const normalized = value.replace(/\r\n/g, "\n");
  if (normalized.length <= limit) return [normalized, false];
  return [normalized.slice(0, limit - 3) + "...", true];
}

function summarizeInputField(
  key: string,
  value: unknown,
): Record<string, unknown> {
  if (typeof value === "string") {
    const [preview, truncated] = truncateText(value);
    if (!truncated) return { [key]: preview };
    return {
      [`${key}_preview`]: preview,
      [`${key}_length`]: value.length,
      [`${key}_truncated`]: true,
    };
  }

  if (
    typeof value === "number" ||
    typeof value === "boolean" ||
    value === null ||
    value === undefined
  ) {
    return { [key]: value };
  }

  if (Array.isArray(value)) {
    const sample: unknown[] = [];
    for (const item of value.slice(0, 3)) {
      if (typeof item === "string") {
        sample.push(truncateText(item, 80)[0]);
      } else if (
        typeof item === "number" ||
        typeof item === "boolean" ||
        item === null ||
        item === undefined
      ) {
        sample.push(item);
      } else if (
        typeof item === "object" &&
        item !== null &&
        !Array.isArray(item)
      ) {
        sample.push({
          keys: Object.keys(item as Record<string, unknown>)
            .sort()
            .slice(0, 8),
        });
      } else {
        sample.push(typeof item);
      }
    }
    const summary: Record<string, unknown> = {
      [`${key}_count`]: value.length,
    };
    if (sample.length > 0) {
      summary[`${key}_sample`] = sample;
    }
    return summary;
  }

  if (typeof value === "object" && value !== null) {
    return {
      [`${key}_keys`]: Object.keys(value as Record<string, unknown>).sort(),
    };
  }

  return { [`${key}_type`]: typeof value };
}

export function summarizeToolUses(
  toolUses: Record<string, unknown>[],
): Record<string, unknown>[] {
  const summarized: Record<string, unknown>[] = [];
  for (const item of toolUses) {
    const toolInput = item.input;
    const summary: Record<string, unknown> = {
      id: item.id,
      name: item.name,
      input: {},
    };
    if (
      typeof toolInput === "object" &&
      toolInput !== null &&
      !Array.isArray(toolInput)
    ) {
      const summarizedInput: Record<string, unknown> = {};
      for (const [key, value] of Object.entries(
        toolInput as Record<string, unknown>,
      )) {
        Object.assign(summarizedInput, summarizeInputField(key, value));
      }
      summary.input = summarizedInput;
    } else {
      summary.input = summarizeInputField("value", toolInput);
    }
    summarized.push(summary);
  }
  return summarized;
}
